File: apps/vsc-plugin/__tests__/build-install-command.test.ts
import { describe, test, expect } from 'vitest';
import { buildInstallAndRun } from '../src/utils/build-install-command';

/**
 * Cross-OS coverage for the install-then-run one-liner the
 * "Pair from mobile" / "Link agent" flows send to the freshly-opened
 * IDE terminal. The job runs on ubuntu-latest AND windows-latest
 * (see `.github/workflows/ci.yml`'s `vsc-plugin` matrix), but the
 * pure function takes the platform + shell as inputs so a single
 * runner exercises every combination deterministically.
 *
 * Background — what we're guarding against (v2.17.x regression):
 *
 *   `npm install -g codeam-cli@latest && codeam pair || npx -y …`
 *
 * `&&` and `||` as pipeline-chain operators were only added in
 * PowerShell 7. PowerShell 5.x — the default shell on every fresh
 * Windows 10/11 install, and the one VS Code picks up via
 * `vscode.env.shell` when the user hasn't installed PS 7 — parses
 * `&&` as an "invalid statement separator" and the command dies
 * before npm runs. We caught it on a Windows Pair From Mobile flow:
 * the terminal opened and the user saw a PowerShell ParserError
 * instead of a paired session.
 *
 * The fix detects PS 5.x specifically (path ends in
 * `…WindowsPowerShell\v1.0\powershell.exe`, NOT pwsh.exe) and emits
 * the semicolon + `$LASTEXITCODE` equivalent that works in PS 5.x
 * AND PS 7+. The tests below pin every relevant shell so any future
 * tweak to the detection / emit logic that drops PS 5.x support is
 * caught at CI time, before a tag.
 */
describe('buildInstallAndRun', () => {
  const REAL_PS5_PATH =
    'C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\powershell.exe';
  const REAL_PS7_PATH = 'C:\\Program Files\\PowerShell\\7\\pwsh.exe';
  const REAL_CMD_PATH = 'C:\\WINDOWS\\System32\\cmd.exe';
  const REAL_BASH_PATH = '/bin/bash';
  const REAL_ZSH_PATH = '/bin/zsh';

  describe('Windows', () => {
    test('PowerShell 5.x — emits semicolon + $LASTEXITCODE (avoids the && parse error)', () => {
      const out = buildInstallAndRun('pair', REAL_PS5_PATH, true);
      expect(out).toContain('npm install -g codeam-cli@latest;');
      expect(out).toContain('if ($LASTEXITCODE -eq 0)');
      expect(out).toContain('codeam pair');
      expect(out).toContain('npx -y codeam-cli@latest pair');
      // The whole command must not use && / || on PS 5.x.
      expect(out).not.toContain('&&');
      expect(out).not.toContain('||');
    });

    test('PowerShell 5.x — detection is case-insensitive (Windows paths)', () => {
      const out = buildInstallAndRun(
        'pair',
        REAL_PS5_PATH.toUpperCase(),
        true,
      );
      expect(out).toContain('if ($LASTEXITCODE -eq 0)');
      expect(out).not.toContain('&&');
    });

    test('PowerShell 7+ (pwsh) — keeps the && / || one-liner', () => {
      const out = buildInstallAndRun('pair', REAL_PS7_PATH, true);
      expect(out).toBe(
        'npm install -g codeam-cli@latest && codeam pair || npx -y codeam-cli@latest pair',
      );
    });

    test('cmd.exe — keeps the && / || one-liner (cmd.exe supports both)', () => {
      const out = buildInstallAndRun('pair', REAL_CMD_PATH, true);
      expect(out).toBe(
        'npm install -g codeam-cli@latest && codeam pair || npx -y codeam-cli@latest pair',
      );
    });

    test('Unknown shell on Windows — falls back to && / || (last-resort same as Unix)', () => {
      const out = buildInstallAndRun('pair', '', true);
      expect(out).toContain('&&');
      expect(out).toContain('||');
    });
  });

  describe('macOS / Linux', () => {
    test('bash — keeps the && / || one-liner', () => {
      const out = buildInstallAndRun('pair', REAL_BASH_PATH, false);
      expect(out).toBe(
        'npm install -g codeam-cli@latest && codeam pair || npx -y codeam-cli@latest pair',
      );
    });

    test('zsh — keeps the && / || one-liner', () => {
      const out = buildInstallAndRun('pair', REAL_ZSH_PATH, false);
      expect(out).toBe(
        'npm install -g codeam-cli@latest && codeam pair || npx -y codeam-cli@latest pair',
      );
    });

    test('shellHint with PS5-shaped suffix but isWindows=false — does NOT trigger PS5 branch', () => {
      // Defence-in-depth: even if a user's `vscode.env.shell` returned
      // something that ended in `powershell.exe` on a Mac (impossible
      // in practice, but the function only knows what we feed it),
      // the macOS terminal would NOT run PowerShell syntax. The
      // `isWindows` gate is the primary filter.
      const out = buildInstallAndRun(
        'pair',
        '/opt/weird/path/powershell.exe',
        false,
      );
      expect(out).toContain('&&');
    });
  });

  describe('subcommand pass-through', () => {
    test('link claude — both branches embed the verb', () => {
      const win = buildInstallAndRun('link claude', REAL_PS5_PATH, true);
      expect(win).toContain('codeam link claude');
      expect(win).toContain('npx -y codeam-cli@latest link claude');

      const unix = buildInstallAndRun('link claude', REAL_BASH_PATH, false);
      expect(unix).toContain('codeam link claude');
      expect(unix).toContain('npx -y codeam-cli@latest link claude');
    });

    test('link codex — same shape, different agent', () => {
      const out = buildInstallAndRun('link codex', REAL_PS5_PATH, true);
      expect(out).toContain('codeam link codex');
      expect(out).toContain('npx -y codeam-cli@latest link codex');
    });
  });
});

File: apps/jetbrains-plugin/src/test/kotlin/com/windsurf/controller/util/BuildInstallCommandTest.kt
package com.windsurf.controller.util

import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * Cross-OS coverage for the install-then-run one-liner the
 * "Pair from mobile" / "Link agent" flows send to the freshly-opened
 * JetBrains terminal.
 *
 * Background — what we're guarding against (v2.17.x regression):
 *
 *     npm install -g codeam-cli@latest && codeam pair || npx -y …
 *
 * `&&` and `||` as pipeline-chain operators were only added in
 * PowerShell 7. PowerShell 5.x — the default shell on every fresh
 * Windows 10/11 install — parses `&&` as an "invalid statement
 * separator" and the command dies before npm runs. JetBrains
 * terminals on Windows default to PowerShell, so a tap on "Pair"
 * from mobile would show the user a PowerShell ParserError instead
 * of a paired session.
 *
 * The fix branches on `os.name`: Windows gets the semicolon +
 * `$LASTEXITCODE` form (works in PS 5.x AND PS 7+); macOS / Linux /
 * WSL keep the original `&&`/`||` form. These tests pin every branch
 * so any future tweak that drops PS 5.x compatibility is caught at
 * CI time, before a tag.
 */
class BuildInstallCommandTest {

    @Test
    fun `Windows 10 emits semicolon + LASTEXITCODE form (no && or ||)`() {
        val cmd = BuildInstallCommand.forSubcommand("pair", "Windows 10")
        assertTrue(cmd.contains("npm install -g codeam-cli@latest;"), cmd)
        assertTrue(cmd.contains("\$LASTEXITCODE"), cmd)
        assertTrue(cmd.contains("codeam pair"), cmd)
        assertTrue(cmd.contains("npx -y codeam-cli@latest pair"), cmd)
        // Critical: PS 5.x can't parse these.
        assertFalse(cmd.contains("&&"), "Windows command must not use && (PS 5.x fails to parse it)")
        assertFalse(cmd.contains("||"), "Windows command must not use || (PS 5.x fails to parse it)")
    }

    @Test
    fun `Windows 11 picks the same branch (case-insensitive os name match)`() {
        val cmd = BuildInstallCommand.forSubcommand("pair", "Windows 11")
        assertTrue(cmd.contains("\$LASTEXITCODE"), cmd)
        assertFalse(cmd.contains("&&"))
    }

    @Test
    fun `Windows Server 2022 also routes to PowerShell-safe form`() {
        val cmd = BuildInstallCommand.forSubcommand("pair", "Windows Server 2022")
        assertTrue(cmd.contains("\$LASTEXITCODE"), cmd)
        assertFalse(cmd.contains("&&"))
    }

    @Test
    fun `os name lowercased still detects Windows (defence-in-depth)`() {
        // os.name doesn't actually come out lower-cased on Windows
        // — JVM always returns "Windows 10" / "Windows 11" — but the
        // `startsWith("Windows", ignoreCase = true)` check should
        // tolerate any variant in case a non-Oracle JVM ever differs.
        val cmd = BuildInstallCommand.forSubcommand("pair", "windows 11")
        assertTrue(cmd.contains("\$LASTEXITCODE"))
    }

    @Test
    fun `macOS keeps amp-amp and pipe-pipe one-liner`() {
        val cmd = BuildInstallCommand.forSubcommand("pair", "Mac OS X")
        assertEquals(
            "npm install -g codeam-cli@latest && codeam pair || npx -y codeam-cli@latest pair",
            cmd,
        )
    }

    @Test
    fun `Linux keeps amp-amp and pipe-pipe one-liner`() {
        val cmd = BuildInstallCommand.forSubcommand("pair", "Linux")
        assertEquals(
            "npm install -g codeam-cli@latest && codeam pair || npx -y codeam-cli@latest pair",
            cmd,
        )
    }

    @Test
    fun `link codex subcommand passes through on Windows`() {
        val cmd = BuildInstallCommand.forSubcommand("link codex", "Windows 11")
        assertTrue(cmd.contains("codeam link codex"), cmd)
        assertTrue(cmd.contains("npx -y codeam-cli@latest link codex"), cmd)
        assertFalse(cmd.contains("&&"))
    }

    @Test
    fun `link claude subcommand passes through on Linux`() {
        val cmd = BuildInstallCommand.forSubcommand("link claude", "Linux")
        assertEquals(
            "npm install -g codeam-cli@latest && codeam link claude || npx -y codeam-cli@latest link claude",
            cmd,
        )
    }
}
